cmd/broker-juggler/main.go

```diff
@@ -0,0 +1,253 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"time"
+
+	"github.com/kelseyhightower/envconfig"
+	"gopkg.in/yaml.v2"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	eventingclient "knative.dev/eventing/pkg/client/injection/client"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/injection/clients/dynamicclient"
+)
+
+type envConfig struct {
+	SystemNamespace string `envconfig:"SYSTEM_NAMESPACE" default:"knative-eventing"`
+	DryRun          bool   `envconfig:"DRY_RUN" default:"false"`
+}
+
+func main() {
+	ctx := injectionEnabled()
+
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		log.Printf("[ERROR] Failed to process env var: %s", err)
+		os.Exit(1)
+	}
+
+	k8s := kubeclient.Get(ctx)
+	client := eventingclient.Get(ctx)
+
+	nss, err := k8s.CoreV1().Namespaces().List(metav1.ListOptions{})
+	if err != nil {
+		panic(err)
+	}
+
+	var cleanups []corev1.ObjectReference
+
+	for i := 0; i < 16; i++ {
+		// Reset.
+		cleanups = make([]corev1.ObjectReference, 0)
+
+		interrupted := false
+
+		for _, name := range []string{"broker-controller", "broker-filter", "broker-ingress"} {
+
+			if d, err := k8s.AppsV1().Deployments(env.SystemNamespace).Get(name, metav1.GetOptions{}); err != nil {
+				fmt.Printf("# [error] %s\n", err)
+			} else if isScaledZero(d) {
+				d.Kind = "Deployment"
+				d.APIVersion = "apps/v1"
+
+				fmt.Printf("---\n\n%s\n\n", toYaml(d))
+
+				cleanups = append(cleanups, corev1.ObjectReference{
+					Kind:       d.Kind,
+					Namespace:  d.Namespace,
+					Name:       d.Name,
+					APIVersion: d.APIVersion,
+				})
+			} else {
+				fmt.Printf("#  Waiting until system namespace broker components are scaled to zero. [try %d/16]\n", i+1)
+				interrupted = true
+				break
+			}
+		}
+
+		if interrupted {
+			if i == 15 {
+				panic("control plane is not in the expected state, is the mtbroker installed?")
+			}
+			time.Sleep(time.Duration(math.Exp2(float64(i))) * time.Second)
+		} else {
+			break
+		}
+	}
+
+	for _, ns := range nss.Items {
+		brokers, err := client.EventingV1beta1().Brokers(ns.Name).List(metav1.ListOptions{})
+		if err != nil {
+			fmt.Printf("# [error] %s\n", err)
+		}
+
+		foundBrokerForCleaning := false
+		for _, broker := range brokers.Items {
+			clean := false
+
+			if broker.Annotations["eventing.knative.dev/broker.class"] == "ChannelBasedBroker" {
```

Comment:
Just jotting this down here so I don't forget. Should we patch the Broker to MTChannelBasedBroker here so it doesn't get orphaned?